### PR TITLE
Feat/publish rework

### LINKED ISSUE #40: [BUG] setDatatokenNameAndSymbol should be part of the service builder
In the for every asset is every datatoken mapped to a specific service (multiple datatokens per asset are possible):

```json
    "datatokens": [
      {
        "address": "0x6fD17dF555881cBA2c2686fB5F71Bb38f5aA132d",
        "name": "deltaDAO AlgoToken 001",
        "serviceId": "06ca5b39d4febb672046c42348935c957804d38c918b3e47c4b3fd8311e8b558",
        "symbol": "DD-AT-001"
      }
    ],
```

- [ ] check if this is also wrong in ocean.js
- [ ] move `setDatatokenNameAndSymbol` from `assetBuilder` to `serviceBuilder`

### PR DESCRIPTION
**Pull request by moritzkirstein**
Introduces a complete overhaul to the asset publishing process.

Implements #29 
Implements #32  
Implements #30  
Fixes #40   

## Proposed Changes

  - Introduce `NautilusDDO` wrapper class
  - Update `NautilusService` and `NautilusAsset` classes respectively
  - Rework the publishing flow, using a multi-step approach to
    1. Create a NFT asset
    2. Create Datatokens and Pricings for all Services of the offering
    3. Create a valid DDO with previous tsx
    4. Publish the DDO on the NFT asset
  - Introduce and improve unit tests for `NautilusDDO` and `NautilusService`

## Additional Information
Ignore `/docs/*` for review, as these are automatically generated on release. Documentation needs to be updated before 0.3.0 release.
Some tests need to be re-enabled before release.

### CODE REVIEW

File: src/publish/index.ts [hunk 1]
```diff
@@ -1,99 +1,134 @@
...
+  const nftAddress = await nftFactory.createNFT(publisherAccount, nftParams)
+
+  LoggerInstance.debug('[publish] NFT published:', nftAddress)
+  return { nftAddress }
+}
+
+export async function createDatatokenAndPricing(config: CreateDatatokenConfig) {
   // --------------------------------------------------
-  // 1. Create NFT & datatokens & create pricing schema
+  // 1. Create Datatoken
```
Comment (Abrom8) [inline src/publish/index.ts:35]: nitpick: Step 1 duplicate

File: src/publish/index.ts [hunk 2]
```diff
@@ -1,99 +1,134 @@
...
+    datatokenParams.name,
+    datatokenParams.symbol,
+    datatokenParams.templateIndex
+  )
+
+  LoggerInstance.debug('[publish] Datatoken created.', datatokenAddress)
 
   // --------------------------------------------------
-  // 2. Construct and encrypt DDO
+  // 2. Create Pricing
```
Comment (Abrom8) [inline src/publish/index.ts:60]: nitpick: step 3

File: src/Nautilus/Asset/NautilusDDO.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    nautilusDDO.id = ddo.id
+    nautilusDDO.context = ddo['@context']
+    nautilusDDO.nftAddress = ddo.nftAddress
+    nautilusDDO.chainId = ddo.chainId
+    nautilusDDO.version = ddo.version
+
+    return nautilusDDO
+  }
+
+  private async buildDDOServices(encrypt = true): Promise<Service[]> {
```
[the remark this answers is not in the document]
Comment (moritzkirstein) [reply src/Nautilus/Asset/NautilusDDO.ts:34]: Overlooked this, logic was changed, so this should no longer be needed

File: src/Nautilus/Asset/NautilusDDO.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+      services: newServices
+    }
+
+    return this.ddo
+  }
+
+  async getDDO(
+    create = false,
+    chainId?: number,
+    nftAddress?: string
```
[the remark this answers is not in the document]
Comment (moritzkirstein) [reply src/Nautilus/Asset/NautilusDDO.ts:136]: Agreed. Especially to stay in line with rest of repo. I'll update this.

**Review by Abrom8 — APPROVED** [review]

**Pull request merged**